vsintegration/src/FSharp.Editor/Refactor/AddExplicitReturnType.fs

```diff
@@ -0,0 +1,125 @@
+open Microsoft.CodeAnalysis.CodeRefactorings
+open Microsoft.CodeAnalysis.CodeActions
+open CancellableTasks
+
+[<ExportCodeRefactoringProvider(FSharpConstants.FSharpLanguageName, Name = "AddExplicitReturnType"); Shared>]
+type internal AddExplicitReturnType [<ImportingConstructor>] () =
+    inherit CodeRefactoringProvider()
+
+    static member isValidMethodWithoutTypeAnnotation
+        (symbolUse: FSharpSymbolUse)
+        (parseFileResults: FSharpParseFileResults)
+        (funcOrValue: FSharpMemberOrFunctionOrValue)
+        =
+        let typeAnnotationRange =
+            parseFileResults.TryRangeOfReturnTypeHint(symbolUse.Range.Start, false)
+
+        let isLambdaIfFunction =
+            funcOrValue.IsFunction
+            && parseFileResults.IsBindingALambdaAtPosition symbolUse.Range.Start
+
+        let res =
+            (not funcOrValue.IsValue || not isLambdaIfFunction)
+            && not (funcOrValue.ReturnParameter.Type.IsUnresolved)
+            && not (parseFileResults.IsTypeAnnotationGivenAtPosition symbolUse.Range.Start)
+            && not typeAnnotationRange.IsNone
+
+        match res with
+        | true -> Some funcOrValue
+        | false -> None
+
+    static member refactor
+        (context: CodeRefactoringContext)
+        (symbolUse: FSharpSymbolUse, memberFunc: FSharpMemberOrFunctionOrValue, parseFileResults: FSharpParseFileResults)
+        =
+        let title = SR.AddExplicitReturnTypeAnnotation()
+
+        let getChangedText (sourceText: SourceText) =
+            let inferredType = memberFunc.ReturnParameter.Type.TypeDefinition.DisplayName
+
+            let rangeOfReturnType =
+                parseFileResults.TryRangeOfReturnTypeHint(symbolUse.Range.Start, false)
+
+            let textChange =
+                rangeOfReturnType
+                |> Option.map (fun range -> RoslynHelpers.FSharpRangeToTextSpan(sourceText, range))
+                |> Option.map (fun textSpan -> TextChange(textSpan, $":{inferredType}"))
+
+            match textChange with
+            | Some textChange -> sourceText.WithChanges(textChange)
+            | None -> sourceText
+
+        let codeActionFunc: CancellationToken -> Task<Document> =
+            fun (cancellationToken: CancellationToken) ->
+                task {
+                    let! sourceText = context.Document.GetTextAsync(cancellationToken)
+                    let changedText = getChangedText sourceText
+
+                    let newDocument = context.Document.WithText(changedText)
+                    return newDocument
+                }
+
+        let codeAction = CodeAction.Create(title, codeActionFunc, title)
+
+        do context.RegisterRefactoring(codeAction)
+
+    static member ofFSharpMemberOrFunctionOrValue(symbol: FSharpSymbol) =
+        match symbol with
+        | :? FSharpMemberOrFunctionOrValue as v -> Some v
+        | _ -> None
+
+    override _.ComputeRefactoringsAsync context =
+        let ct = context.CancellationToken
```

Comment:
Currently it the codebase it is probably a bit more common to get the token from the cancellable task itself:
`CancellableTask.getCancellationToken ()`

Reply:
Huh.. I think I have a knot in my brain with this one... getCancellationToken gives me a partially applied method that still excpects a cancellation token. How do I use this with CancellableTask.startWithTask? or is that the wrong method

Reply:
Well technically it should work just like `let! cancellationToken = CancellableTask.getCancellationToken ()`.
E.g. [here](https://github.com/dotnet/fsharp/blob/main/vsintegration/src/FSharp.Editor/CodeFixes/AddMissingFunKeyword.fs#L52). This doesn't compile for you here?

Reply:
The method signature is different. I need a task here. I could just omit the handling of cancellationTokens, see current version

Reply:
That looks better :) I will take a deeper look soon and see if something can be simplified here but for now that's okay.